kedro/io/data_catalog.py

```diff
@@ -436,27 +439,33 @@ def __contains__(self, data_set_name):
             return True
         return False
 
+    @classmethod
     def _resolve_config(
-        self,
+        cls,
         data_set_name: str,
         matched_pattern: str,
+        config: dict,
     ) -> dict[str, Any]:
         """Get resolved AbstractDataset from a factory config"""
         result = parse(matched_pattern, data_set_name)
-        config_copy = copy.deepcopy(self._dataset_patterns[matched_pattern])
         # Resolve the factory config for the dataset
-        for key, value in config_copy.items():
-            if isinstance(value, Iterable) and "}" in value:
-                # result.named: gives access to all dict items in the match result.
-                # format_map fills in dict values into a string with {...} placeholders
-                # of the same key name.
-                try:
-                    config_copy[key] = str(value).format_map(result.named)
-                except KeyError as exc:
-                    raise DatasetError(
-                        f"Unable to resolve '{key}' for the pattern '{matched_pattern}'"
-                    ) from exc
-        return config_copy
+        if isinstance(config, dict):
+            for key, value in config.items():
+                config[key] = cls._resolve_config(data_set_name, matched_pattern, value)
+        elif isinstance(config, list):
+            config = [
+                cls._resolve_config(data_set_name, matched_pattern, value)
+                for value in config
+            ]
+        elif isinstance(config, Iterable) and "}" in config:
+            try:
```

Comment:
Why list and Iterable are treated differently? You mentioned there is a new example from user maybe it worth to document it in the PR/issue.

Reply:
Hi @noklam, yes I provided an example in #2992

Iterable here basically means string because of the second condition. 
In fact, I think having `and "}" in config:` is actually misleading if you expect the condition to work with nested dictionaries and lists. 

Indeed, while strings, lists and dicts are Iterable, the second check fails for dict and list, see example below

```python
from typing import Iterable

config = "my{param}" 
print(isinstance(config, Iterable)) # True
print("}" in config) # True

config = ["my{param}"]
print(isinstance(config, Iterable)) # True
print("}" in config) # False

config = {"mykey":"my{param}"}
print(isinstance(config, Iterable)) # True
print("}" in config) # False
```

Which is why the solution is to only go inside the base case when you actually substitute a string. 
To make things clearer, maybe `Iterable` should be replaced by `str`.

Reply:
Yep this should be string

Reply:
Thanks for explaining! It make perfect sense and I agree it makes more sense with `str`.